sdk/cxx/src/message_dispatcher.cpp

```diff
@@ -0,0 +1,146 @@
+/*
+ Copyright 2017 Intel Corporation
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+------------------------------------------------------------------------------
+*/
+
+#include <exception>
+#include <iostream>
+#include <string>
+#include <thread>
+
+#include <log4cxx/logger.h>
+#include <zmqpp/context.hpp>
+#include <zmqpp/poller.hpp>
+#include <zmqpp/socket.hpp>
+#include <zmqpp/socket_types.hpp>
+#include <zmqpp/socket_options.hpp>
+
+#include "sawtooth/message_dispatcher.h"
+
+
+namespace sawtooth {
+
+static log4cxx::LoggerPtr logger(log4cxx::Logger::getLogger
+    ("sawtooth.MessageDispatcher"));
+
+MessageDispatcher::MessageDispatcher():
+        context_(),
+        server_socket(this->context_, zmqpp::socket_type::dealer),
```

Comment:
If you drink the ZMQ kool-aid, socket creation should be done in the thread where the socket is used, so in `MesageDispatcher::DispatchThread()`.